### PR TITLE
Append HMAC to BitBox messages

### PR DESCRIPTION
**Pull request by TheCharlatan**
As implemented in upstream electrum https://github.com/spesmilo/electrum/pull/4895 messsages to the BitBox are now authenticated by appending an HMAC. Since this uses some encoding functions from upstream electrum, these have now been defined in the digitalbitbox plugin file. They can be moved to their appropriate place in bitcoin.py once upstream changes are ported.

Also includes a small fix to the seed command from upstream https://github.com/spesmilo/electrum/pull/4906 .

### CODE REVIEW

**Comment by cculianu** [comment]
Thanks for this. 

Will review this today/tonight or tomorrow and get back to you.

**Comment by markblundeberg** [comment]
At a quick glance, looks solid. Extra points for using hmac.compare_digest. :+1:

**Comment by cculianu** [comment]
Yeah sorry I still haven't had a chance to review this -- I will do so.. be patient.  Some stuff came up.

File: plugins/digitalbitbox/digitalbitbox.py [hunk 1]
```diff
@@ -275,7 +310,7 @@ def mobile_pairing_dialog(self):
     def dbb_generate_wallet(self):
         key = self.stretch_key(self.password)
         filename = ("Electrum-" + time.strftime("%Y-%m-%d-%H-%M-%S") + ".pdf").encode('utf8')
-        msg = b'{"seed":{"source": "create", "key": "%s", "filename": "%s", "entropy": "%s"}}' % (key, filename, b'Digital Bitbox Electrum Plugin')
+        msg = b'{"seed":{"source": "create", "key": "%s", "filename": "%s", "entropy": "%s"}}' % (key, filename, to_hexstr(os.urandom(32)))
```
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: I like the salt.

File: plugins/digitalbitbox/digitalbitbox.py [hunk 2]
```diff
@@ -39,6 +41,32 @@
...
+    else:
+        return hmac.new(key, msg, digest).digest()
+
+def EncodeAES_bytes(secret: bytes, msg: bytes) -> bytes:
+    assert_bytes(msg)
+    iv = bytes(os.urandom(16))
+    ct = aes_encrypt_with_iv(secret, iv, msg)
+    return iv + ct
+
+def DecodeAES_bytes(secret: bytes, ciphertext: bytes) -> bytes:
```
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: Heh, looks good.  Makes me wonder if we should offer this more generic funciton and have our own `bitcoin.EncodeAES` & `bitcoin.DecodeAES` call the byte-level one...

File: plugins/digitalbitbox/digitalbitbox.py [hunk 3]
```diff
@@ -39,6 +41,32 @@
 def to_hexstr(s):
     return binascii.hexlify(s).decode('ascii')
 
+def derive_keys(x):
+    h = Hash(x)
+    h = hashlib.sha512(h).digest()
+    return (h[:32],h[32:])
+
+def hmac_oneshot(key: bytes, msg: bytes, digest) -> bytes:
```
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: Hmm. Perhaps I need to think about putting this one in bitcoin.py.

**Comment by cculianu** [comment]
Ok.. I did the review. Looks great!  I lack a bitbox to verify but I presume you tested it and it works, right @TheCharlatan ?

Thanks so much for submitting this PR!

Let me ponder whether to port the upstream changes (Likely I will do so).

In that case you  may have to force-push your reduced version...

Will get back to you soon.

File: plugins/digitalbitbox/digitalbitbox.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 #
 
 try:
-    from electroncash.bitcoin import TYPE_ADDRESS, push_script, var_int, msg_magic, Hash, verify_message, pubkey_from_signature, point_to_ser, public_key_to_p2pkh, EncodeAES, DecodeAES, MyVerifyingKey, int_to_hex
+    from electroncash.bitcoin import TYPE_ADDRESS, push_script, var_int, msg_magic, Hash, verify_message, pubkey_from_signature, point_to_ser, public_key_to_p2pkh, EncodeAES, DecodeAES, MyVerifyingKey, int_to_hex, aes_encrypt_with_iv, aes_decrypt_with_iv, assert_bytes
```
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: Ok man, I refactored `bitcoin.py` in master. Please go ahead and push another commit on top of this one (or force push a commit amendment to this one, whatever you prefer) that:

1. imports the new `EncodeAES_bytes`, `DecodeAES_bytes`, `hmac_oneshot`
2. Gets rid of any extant references to the old `EncodeAES` and `DecodeAES` (these are now renamed to `*_base64`)
3. Any other clean-up you see..

Again thanks so much for this.

File: plugins/digitalbitbox/digitalbitbox.py (hunk 3 above)
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: Remove (can not be imported from `bitcoin.py`)

File: plugins/digitalbitbox/digitalbitbox.py [hunk 5]
```diff
@@ -39,6 +41,32 @@
...
+    return (h[:32],h[32:])
+
+def hmac_oneshot(key: bytes, msg: bytes, digest) -> bytes:
+    if hasattr(hmac, 'digest'):
+        # requires python 3.7+; faster
+        return hmac.digest(key, msg, digest)
+    else:
+        return hmac.new(key, msg, digest).digest()
+
+def EncodeAES_bytes(secret: bytes, msg: bytes) -> bytes:
```
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: Remove

File: plugins/digitalbitbox/digitalbitbox.py (hunk 2 above)
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: rm

**Comment by TheCharlatan** [comment]
Force-pushed. The plugin now imports hmac_oneshot, EncodeAES_bytes and DecodeAES_bytes from bitcoin.py.

File: plugins/digitalbitbox/digitalbitbox.py [hunk 6]
```diff
@@ -4,7 +4,7 @@
 #
 
 try:
-    from electroncash.bitcoin import TYPE_ADDRESS, push_script, var_int, msg_magic, Hash, verify_message, pubkey_from_signature, point_to_ser, public_key_to_p2pkh, EncodeAES, DecodeAES, MyVerifyingKey, int_to_hex
+    from electroncash.bitcoin import TYPE_ADDRESS, push_script, var_int, msg_magic, Hash, verify_message, pubkey_from_signature, point_to_ser, public_key_to_p2pkh, EncodeAES, DecodeAES, MyVerifyingKey, int_to_hex, hmac_oneshot, EncodeAES_bytes, DecodeAES_bytes
```
Comment (cculianu) [inline plugins/digitalbitbox/digitalbitbox.py]: Would you mind terribly removing the import for `EncodeAES` and `DecodeAES`?  I'm trying to deprecate those functions to be in-line with the upstream APIs. :)

If you're too busy let me know and I'll just merge as-is and change it myself later..

**Comment by cculianu** [comment]
Hey man.  You're The Man!  Thank you. :)

1 small nit -- I wrote a comment.  if you can get to it great, if not I'll do it.

Cheers and thanks a lot, again, for this PR.

**Comment by TheCharlatan** [comment]
Force-pushed. Removed import of EncodeAES and changed to EncodeAES_base64.

**Comment by cculianu** [comment]
Ah, great.  Now I think I can get rid of this API! (This was the only plugin using it, as far as I know.. ha ha.) Thanks man!

**Pull request merged**